contrib/ircdwatch/ircdwatch.8
.\" @(#)$Id$
.TH IRCDWATCH 8 "$Date$"
.SH NAME
ircdwatch \- Daemon for monitoring ircd and ircd.conf
.SH SYNOPSIS
.hy 0
.IP \fBircdwatch\fP
[
.BI \-\-kill
] [
.BI \-\-rest
] [
.BI \-\-help
]
.SH DESCRIPTION
.LP
\fIircdwatch\fP is a daemon that periodically checks that the 
\fIircd\fP daemon is running, restarting it of necessary.  This
daemon can also be configured (at compile time) to check for
changes to the \fIircd\fP configuration file and make \fIircd\fP
reload its configuration file by sending it a SIGHUP signal.  
.LP
Given command line arguments \fIircdwatch\fP will serve as a remote
control for stopping both \fIircdwatch\fP and \fIircd\fP or just
\fIircdwatch\fP.

.SH OPTIONS
.TP
.B \-\-kill
Stop both \fIircd\fP and \fIircdwatch\fP
.TP
.B \-\-rest
Stop \fIircdwatch\fP but leave \fIircd\fP running
.TP
.B \-\-help
List command line arguments
.SH COPYRIGHT
Copyright (C) 1998 Bjorn Borud, <borud@guardian.no>
.LP
.RE
.SH FILES
 "ircd.conf"
 "ircd.pid"
 "ircdwatch.pid"
.SH "SEE ALSO"
ircd(8)
.SH BUGS
There may be some portability issues.  Report bugs to <ircd-users@irc.org>
.SH AUTHOR
Bjorn Borud, <borud@guardian.no>


doc/RELEASE_NOTES
# @(#)$Id$

This is version 2.11.0 of the IRC software.
===============================================================================
Important changes in 2.11.0 (since 2.10.3p5):

	* New server-server protocol based on Unique IDs. In pure 2.11 collided
	  clients will stay on with changed nick.
	* M: requires SID now. Please read INSTALL.txt and example.conf
	* A: accepts "network name" parameter.
	* H: can have SID masks to accept. Read docs.
	* Nicknames of local clients cannot have [\]{|} chars. Sorry.
	* Channelnames created on 2.11.0 will have chars {}~\ in their name
	  automatically converted to []^|. Sorry.
	* I: require now both name AND addr to match. Revise your config.
	* I: have flags now (D, E, e, F, I, M, N, R), check INSTALL for details.
	* OPER_REMOTE is removed (and server behaves like it was defined).
	* stats l/L no longer returns listeners, use stats P.
	* DNS command is oper only from now.
	* DELAY_CLOSE is default, this may use some extra fds.
	* resolving slightly changed: when host has CNAMEs, use the name that
	  client IP PTR points to, not the CNAME that points to client IP.
	* resolv.conf can have IPv6 nameservers (this makes ircd fully IPv6).
	* P: can have flags ('D' and 'S'), check INSTALL file for details.
	* class usage was moved from end of trace to stats Y.
	* redundant bans are allowed to be set now.
	* SERVICE command has been changed slightly (removed 2nd (stok) and 5th
	  (hopcount) parameters), please update your services.
	* localstatedir in configure changed from var/run to var, be careful.
	* user mode +w is doing nothing: class usage is in stats Y and wallops
	  are sent to a new server channel &WALLOPS.
	* irc client is no longer included.
	* rehash q was changed to rehash d (dns flush).
	* ircd is capable of including files to config file without need of m4.
	* new commands: TKLINE and UNTKLINE.
	* stats K returns list of klines, stats k list of tklines.
	* rehash t removes all tklines.
	* O: have flags now, check INSTALL file for details; WARNING: no flags
	  means no privileges at all!
	* oper-only (access granted via O: flag) &CLIENTS channel reporting
	  (if defined) clients connects, quits, nickchanges.
	* new way of excluding users, X:line.
	* new wildcard (#), matching any digit.

This is version 2.10.3 of the IRC software.
===============================================================================

Important changes in 2.10.3p4 (since 2.10.3p2):
       * ip6.arpa is now default, with defining SIXBONE_HACK we
         also recognize ip6.int
       * RESTRICT_HOSTNAMES is default on, allowing client hosts in
         conformance with RFC 1034, 1035, and 1123 (no '&', '*', '\' or
         other "strange" characters)
       * CLONE_CHECK default on, not SLOW_ACCEPT
       * find_kill() is aware of '+' prefix, so no need to add another
         K:line for it (or using *username)

===============================================================================

Important changes in 2.10.3p2 (since 2.10.3):
	* u@h limits in Y:lines now default to 1.1, not 0.0
	* WHO ** is not special anymore (may break clients depending on it)
	* CIDR support in O:lines
	* no need to use :ffff: mapped IPv4 addresses anymore
	* now ircd reopens logfiles only upon HUP; this means it is not enough
	  to remove (or rename) logfile to stop logging (but we gain a lot!)

===============================================================================

New features in 2.10.3:
	* new options for iauth.conf to better control iauth behaviour
	* iauth now supports dynamically shared modules.
	* socks module now checks for both v4 and v5 of the SOCKS protocol.
	* iauth has a new module: LHEx, see ftp://ftp.irc.org/irc/server/LHEx

Important changes in 2.10.3 (since 2.10.2):
	* default PATHs have changed, see INSTALL file and Makefile.
	* V line code was fixed, and format slightly changed again.

------

Because of the many changes concerning iauth, it is recommended that this new
version of the iauth program not be used with older version of the IRC daemon.

===============================================================================

Version 2.10.2 of the software adds support for IPv6.

Important changes in 2.10.2 (since 2.10.1):
	* iauth's socks module now uses an internal cache.
	* iauth's socks module now checks for SOCKSv4 (rather than v5) proxies.

===============================================================================

2.10 uses a new (server-server) protocol.

New features in 2.10.0:

	* slave process handles authentication (ident lookups, ..).
	* creation of a collision proof type of channels (prefix !).
	* opless !channels may be reoped by the server (mode +r).
	* added channel mode +e (EFnet's exceptions to bans).
	* added channel mode +I (invitations).
	* /invite can now be used to override channel bans & limit.
	* away status is propagated again. (away messages are not).
	* users need +o (or +v) to speak on a channel where they're banned.

Important configuration changes between 2.9.x and 2.10.x:

	* The V line format has changed!

------

If the irc daemon is unable to bind any socket to listen to for incoming
connections, it will die rather than stay alive.

===============================================================================

New feature in 2.9.5:
	* D lines created.

------

2.9.5 is taking steps to suppress the usage of the 2.9 JOIN
format (:nickname JOIN #channel^Gov).  Future versions will
not generate such joins anymore.  In order to make the
transition smooth, it is imperative that all servers on the
IRC network be upgraded to 2.9.5 when the JOIN syntax is
abandonned.  Not doing so will result in a considerable
increase of the amount of bandwidth used during netjoins.

As a result, MIRC_KLUDGE is now defined by default in config.h

------

2.9.5 can be compiled on a W32 system using the Cygwin32
library (http://www.cygnus.com/misc/gnu-win32/).


contrib/ircdwatch/ircdwatch.c
/*
 *   Copyright (C) 1998 Bjorn Borud <borud@guardian.no>
 * 
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 1, or (at your option)
 *   any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

static const volatile char rcsid[] = "@(#)$Id$";

#include <stdio.h>
#include <stdlib.h>     /* atol() */
#include <unistd.h>     /* fork() exec() */
#include <sys/types.h>
#include <sys/stat.h>   /* stat() */
#include <signal.h>
#include <syslog.h>
#include <string.h>     /* strncmp() */
#include <time.h>

#include "os.h"
#include "config.h"

/* 
 * Try and find the correct name to use with getrlimit() for setting
 * the max.  number of files allowed to be open by this process. 
 * (borrowed from ircd/s_bsd.c)
 */
#ifdef RLIMIT_FDMAX
# define RLIMIT_FD_MAX   RLIMIT_FDMAX
#else
# ifdef RLIMIT_NOFILE
#  define RLIMIT_FD_MAX RLIMIT_NOFILE
# else
#  ifdef RLIMIT_OPEN_MAX
#   define RLIMIT_FD_MAX RLIMIT_OPEN_MAX
#  else
#   undef RLIMIT_FD_MAX
#  endif
# endif
#endif

#define PID_LEN 7 /* overkill, but hey */
#define MAX_OPEN_FILEDESCRIPTORS_WILD_GUESS 256

static int want_to_quit = 0;

static void finalize(int i)
{
#ifdef IRCDWATCH_USE_SYSLOG
  syslog(LOG_NOTICE, "ircdwatch daemon exiting");
  closelog();
#endif
  exit(i);
}

static int daemonize(void)
{
  pid_t pid;
  int   i;
  int   open_max;

#ifdef RLIMIT_FD_MAX
  struct rlimit rlim;

  if (getrlimit(RLIMIT_FD_MAX, &rlim) != 0) {
    perror("getrlimit");
    finalize(1);
  }

  /* if we get a lame answer we just take a wild guess */
  if (rlim.rlim_max == 0) {
    open_max = MAX_OPEN_FILEDESCRIPTORS_WILD_GUESS;
  }

  open_max = rlim.rlim_max;

#else

  /* no RLIMIT_FD_MAX?  take a wild guess */
  open_max = MAX_OPEN_FILEDESCRIPTORS_WILD_GUESS;

#endif

  pid = fork();
  if (pid < 0) {
    perror("fork");
    exit(1);
  }

  /* parent process dies */
  if (pid != 0) {
    exit(0);
  }

  setsid();
  umask(0);

  /* close file descriptors */
  for (i=0; i < open_max; i++) {
    close(i);
  }

  return(0);
}

static void sig_handler (int signo) 
{
  if (signo == SIGHUP) {
    want_to_quit = 1;
    return;
  }

  if (signo == SIGALRM) {

#ifndef POSIX_SIGNALS
    (void)signal(SIGALRM, &sig_handler);
#endif

    return;
  }
}


static void set_up_signals(void) 
{
#ifdef POSIX_SIGNALS
  struct sigaction act;

  act.sa_handler = sig_handler;
  sigemptyset(&act.sa_mask);
  act.sa_flags = 0;

  if (sigaction(SIGHUP, &act, NULL) < 0) {
    perror("sigaction");
  }

  if (sigaction(SIGALRM, &act, NULL) < 0) {
    perror("sigaction");
  }
#else
  (void)signal(SIGHUP, &sig_handler);
  (void)signal(SIGALRM, &sig_handler);
#endif

  /* ignore it if child processes die */
  signal(SIGCHLD, SIG_IGN);
}

static int write_my_pid(void)
{
  FILE *f;

  f = fopen(IRCDWATCH_PID_FILENAME, "w");
  if (f == NULL) {
    return(-1);
  }

  fprintf(f, "%d\n", getpid());
  fclose(f);

  return(0);
}


static int file_modified(char *s)
{
  struct stat st;
  
  if (stat(s, &st) < 0) {
    return(-1);
  }
  return(st.st_ctime);
}


static int spawn (char *cmd) 
{
  pid_t pid;

  pid = fork();

  if (pid == -1) {
#ifdef IRCDWATCH_USE_SYSLOG
    syslog(LOG_ERR, "spawn() unable to fork, errno=%d", errno);
#endif
    return(-1);
  }

  if (pid == 0) {
    execl("/bin/sh", "sh", "-c", cmd, (char *) NULL);
    _exit(127);
  }
  return(0);
}

static int read_pid(char *pid_filename) 
{
  FILE *f;
  char pidbuf[PID_LEN];
  pid_t pid;

  f = fopen(pid_filename, "r");
  if (f == NULL) {
#ifdef IRCDWATCH_USE_SYSLOG
    syslog(LOG_ERR, "unable to fopen() %s: %s", pid_filename, strerror(errno));
#endif
    return(-1);
  }

  if (fgets(pidbuf, PID_LEN, f) != NULL) {
    pid = atol(pidbuf);
  } else {
#ifdef IRCDWATCH_USE_SYSLOG
    syslog(LOG_ERR, "fgets() %s: %s", pid_filename, strerror(errno));
#endif
    fclose(f);
    return(-1);
  }

  fclose(f);
  return(pid);
}

static int file_exists (char *s)
{
  struct stat st;
  if ((stat(s, &st) < 0) && (errno == ENOENT)) {
    return(0);
  }
  return(1);
}

/* yeah, I'll get around to these in some later version */

static int file_readable (char *s)
{
  return(access(s, R_OK) == 0);
}

static int file_writable (char *s)
{
  return(access(s, W_OK) == 0);
}

static int file_executable (char *s)
{
  int rc;
  rc = (access(IRCD_PATH, X_OK) == 0);
  return rc;
}

static int verify_pid(int pid) 
{
  int res;

  res = kill(pid, 0);
  if (res < 0 && errno == EPERM) {
    fprintf(stderr, "Not process owner\n");
    exit(1);
  }

  return(res == 0);
}

static int ircdwatch_running (void)
{
  int pid;

  if (file_exists(IRCDWATCH_PID_FILENAME)) {
    pid = read_pid(IRCDWATCH_PID_FILENAME);
    if (pid > 0) {
      return(verify_pid(pid) == 1);
    } else {
      return(-1);
    }
  }
  return(0);
}

static int ircd_running (void)
{
  int pid;

  if (file_exists(IRCDPID_PATH)) {
    pid = read_pid(IRCDPID_PATH);
    if (pid > 0) {
      return(verify_pid(pid) == 1);
    } else {
      return(-1);
    }
  }
  return(0);
}

static void hup_ircd (void)
{
  int pid;
  int res;
  
  if (file_exists(IRCDPID_PATH)) {
    pid = read_pid(IRCDPID_PATH);
    if (pid > 0) {
      res = kill(pid, SIGHUP);
      if (res < 0 && errno == EPERM) {
#ifdef IRCDWATCH_USE_SYSLOG
	syslog(LOG_ERR, "not allowed to send SIGHUP to ircd");
#endif
	finalize(1);
      }
    }
  }
}


static void daemon_run (void) 
{
  int i;
#ifdef IRCDWATCH_HUP_ON_CONFIG_CHANGE
  int last_config_time = 0;
#endif

  /* is ircdwatch already running? */
  i = ircdwatch_running();
  if (i == -1) {
    /* unable to open pid file.  wrong user? */
    fprintf(stderr, "ircdwatch pid file exists but is unreadable\n");
    exit(1);
  } else if (i) {
    fprintf(stderr, "ircdwatch is already running\n");    
    exit(0);
  }

  /* is ircd running? */
  i = ircd_running();
  if (i == -1) {
    /* unable to open pid file.  wrong user? */
    fprintf(stderr, "ircdwatch pid file exists but is unreadable\n");
    exit(1);
  } else  if (!i) {
    fprintf(stderr, "ircd not running. attempting to start ircd...\n");
    if (file_exists(IRCD_PATH)) {
      if (file_executable(IRCD_PATH)) {
	spawn(IRCD_PATH);
      } else {
	fprintf(stderr, "%s not executable\n", IRCD_PATH);
	exit(1);
      }
    } else {
      fprintf(stderr, "%s does not exist\n", IRCD_PATH);
      exit(1);      
    }
  }

  set_up_signals();
  closelog();
  /*  daemonize(); */

#ifdef IRCDWATCH_USE_SYSLOG
  openlog(IRCDWATCH_SYSLOG_IDENT, 
	  IRCDWATCH_SYSLOG_OPTIONS, 
	  IRCDWATCH_SYSLOG_FACILITY);
  
  syslog(LOG_NOTICE, "starting ircdwatch daemon");
#endif

  alarm(IRCDWATCH_POLLING_INTERVAL);
  pause();

  while (!want_to_quit) {
    if (! ircd_running() ) {

#ifdef IRCDWATCH_USE_SYSLOG
      syslog(LOG_ERR, "spawning %s", IRCD_PATH);
#endif

      spawn(IRCD_PATH);
    }

#ifdef IRCDWATCH_HUP_ON_CONFIG_CHANGE
    i = file_modified(IRCDCONF_PATH);
    if (i != -1) {

      if (last_config_time == 0) {
	last_config_time = i;
      } 

      else if (i > last_config_time) {
	last_config_time = i;	
	hup_ircd();

#ifdef IRCDWATCH_USE_SYSLOG
	syslog(LOG_NOTICE, "config change, HUPing ircd");
#endif

      }
    }
#endif

    alarm(IRCDWATCH_POLLING_INTERVAL);
    pause();
  }
  return;
}

static void kill_ircd (void)
{
  int pid;
  int res;
  
  if (file_exists(IRCDPID_PATH)) {
    pid = read_pid(IRCDPID_PATH);
    if (pid > 0) {
      res = kill(pid, SIGTERM);
      if (res < 0) {
	perror("ircd kill");
	finalize(1);
      } else {
	fprintf(stderr, "killed ircd\n");
      }
    }
  } else {
    fprintf(stderr, "File %s does not exist\n", IRCDPID_PATH);
  }
}

static void kill_ircdwatch (void)
{
  int pid;
  int res;
  
  if (file_exists(IRCDWATCH_PID_FILENAME)) {
    pid = read_pid(IRCDWATCH_PID_FILENAME);
    if (pid > 0) {
      res = kill(pid, SIGHUP);
      if (res < 0) {
	perror("ircdwatch kill");
	finalize(1);
      } else {
	fprintf(stderr, "Sent HUP to ircdwatch.  it will die soon...\n");
      }
    }
  } else {
    fprintf(stderr, "File %s does not exist\n", IRCDWATCH_PID_FILENAME);
  }
}


static void usage (void)
{
  fprintf(stderr,"\n\
Usage:\n\
  ircdwatch [--kill | --rest | --help]\n\
\n\
     --kill, stop both ircdwatch and ircd\n\
     --rest, stop ircdwatch but let ircd alone\n\
     --help, display this text\n\
\n\
%s\n", rcsid);
}

int
main (int argc, char *argv[])
{

#ifdef IRCDWATCH_USE_SYSLOG
  openlog(IRCDWATCH_SYSLOG_IDENT, 
	  IRCDWATCH_SYSLOG_OPTIONS, 
	  IRCDWATCH_SYSLOG_FACILITY);
#endif

  if (argc > 1) {
    if (strncmp(argv[1], "--rest", 6) == 0) {
      kill_ircdwatch();
      exit(0);
    }

    if (strncmp(argv[1], "--kill", 6) == 0) {
      kill_ircdwatch();
      kill_ircd();
      exit(0);
    }

    usage();
    exit(0);
  }

  daemon_run();
  finalize(0);
  return 0;
}
